### DISCUSSION TITLE
Get nearest lines of points in shapefiles

### DISCUSSION THREAD

**Discussion by forestbat** [geopandas/geopandas#2773] [Q&A] [answered]
[https://github.com/geopandas/geopandas/discussions/2733](url)
Now I have a shapefile which have 364 points，and I want to get 364 linestrings which are the nearest from these points from another shapefile，but when I use this code：
```
# gpd_nodes_df contains 364 nodes, and gpd_lines_df contains 105157 linestrings
# because I want to get 364 linestrings, so I choose 'right' as parameter
gpd_min_line_df = gpd.sjoin_nearest(gpd_nodes_df, gpd_lines_df, 'right')
```
`gpd_min_line_df` has 106959 linestrings.
Because I don't know distances from points to linestrings, so I can't appoint a fixed distance for `sjoin_nearest` method.
How to solve my problem?  Thanks for a lot.

**Comment by martinfleis** [comment, marked as answer]
You need "left" join, not "right". You misunderstood the parameter.

**Reply by forestbat** [reply]
![图片](https://user-images.githubusercontent.com/23413915/216097708-bae5a025-291b-4fc3-bf79-54bb09eb507a.png)
Is it this “index_right” colomn？I tested left join but I ignored it.

**Reply by martinfleis** [reply]
`index_right` represents the values from the index of `gpd_lines_df` here.

**Reply by forestbat** [reply]
But this is another problem：length of `index_right` is 369，and length of my `gpd_nodes_df` is 364，why there is difference？

**Reply by martinfleis** [reply]
You may have some equidistant lines. Without a reproducible example, I will be just guessing.

**Reply by forestbat** [reply]
It's pity that I can’t give you points' shapefiles，but there is a strange thing：4
![图片](https://user-images.githubusercontent.com/23413915/216281864-e21fbf09-62e8-4dd1-9763-442bc6c09212.png)
shape of `gpd_min_line_df` is (369,23), however there is only 364 lines in this dataframe:
![图片](https://user-images.githubusercontent.com/23413915/216280546-4bc0fe3c-9ba1-4ff8-9c2c-2a9da6a4cc7b.png)
And if I try to drop row 364-368 from `gpd_min_line_df`, program tells me these rows don't exist in it.
`KeyError: '[364 365 366 367 368] not found in axis'`
So what happened do you think?

**Reply by martinfleis** [reply]
As I said, you may have some equidistant lines. The index will have duplicates as your points get joined to multiple lines that are of the same distance (so there is not one nearest but two equally near). That is why you don't have values higher that 363 but have lenght of 369.

**Reply by forestbat** [reply]
Yes，as you think there are some equidistant lines，I use this code to debug：
```(python)
for index in range(0, len(gpd_node_df)):
        node_df = gpd.GeoDataFrame(gpd_node_df.iloc[index]).transpose()
        min_line_df: GeoDataFrame = gpd.sjoin_nearest(node_df, gpd_line_df, 'left')
        min_line_df
```
![图片](https://user-images.githubusercontent.com/23413915/216301523-408537bd-995e-4a71-92e7-e93122d60a5a.png)
So I hope you can change code of  `sjoin_nearest` method，make equidistant lines disappear, or you can try to tell me how to solve this problem by myself.

**Reply by martinfleis** [reply]
You can drop the duplicates as 
```py
df = df.loc[~df.index.duplicated()]
```

I believe that the behaviour of `sjoin_nearest` is correct as we do not have a way of telling which of the two shall be dropped.
